awscli/autocomplete/completer.py

```diff
@@ -0,0 +1,120 @@
+            if result is not None:
+                return result
+        return []
+
+
+class BaseCompleter(object):
+    def complete(self, parsed):
+        """Attempt to autocomplete parsed on parsed result.
+
+        Subclasses should implement this method.
+
+        :param parsed: A ParsedResult from the auto complete CLI parser.
+        :return: This method should return one of two values:
+
+            * ``None`` - If the completer doesn't understand how to
+              complete the command, it should return ``None``.  This
+              signals to the ``AutoCompleter`` that it should move on
+              to the next completer.
+            * ``List[str]`` - If the completer is able to offer
+              auto-completions it should return a list of strings that
+              are valid suggestions for completing the command.  This
+              indicates to the ``AutoCompleter`` to immediately return
+              and to stop consulting other completers for results.
+        """
+        raise NotImplementedError("complete")
+
+
+class ModelIndexCompleter(BaseCompleter):
+
+    def __init__(self, index):
+        self._index = index
+
+    def complete(self, parsed):
+        if parsed.unparsed_items or parsed.last_fragment is None:
+            # If there's ever any unparsed items, then the parser
+            # encountered something it didn't understand.  We won't
+            # attempt to auto-complete anything here.
+            return
+        last_fragment = parsed.last_fragment
+        if last_fragment.startswith('--'):
+            # We could technically offer completion of options
+            # if the last fragment is an empty string, but to avoid
+            # dumping too much information back to the user, we only
+            # offer completions for options if the value starts with
+            # '--'.
+            return self._complete_options(parsed)
+        else:
+            return self._complete_command(parsed)
+
+    def _complete_command(self, parsed):
+        lineage = parsed.lineage + [parsed.current_command]
+        result = [name for name in self._index.command_names(lineage)
+                  if name.startswith(parsed.last_fragment)]
+        return result
+
+    def _complete_options(self, parsed):
+        # '--endpoint' -> 'endpoint'
+        fragment = parsed.last_fragment[2:]
+        arg_names = self._index.arg_names(
+            lineage=parsed.lineage, command_name=parsed.current_command)
+        results = [
+            '--%s' % arg_name for arg_name in arg_names
+            if arg_name.startswith(fragment)]
+        # Global params apply to any scope, so if we're not
+        # in the global scope, we need to add completions for
+        # global params
+        self._inject_global_params_if_needed(parsed, results, fragment)
+        return results
+
+    def _inject_global_params_if_needed(self, parsed, results, fragment):
+        is_in_global_scope = (
```

Comment:
I'm still not sure if I am following this. We only want to suggest global arguments if the current command is `aws` and there is no subcommands? So something like `aws --endpoint` would not trigger this if statement? Or is this because it is handled at the beginning of `_complete_options()`? It looks like it does get handled based on the tests.

Reply:
What's happening is that we offer CLI options to auto complete based on your scope, e.g `aws ec2 describe-instances --` will autocomplete based on the params available for the `describe-instances` operation.  When you're in the global scope (`aws --`), the "current scope" *is* the global scope, so the only valid completions in that example would be global params.  However, when you're in the scope of a single operation, in addition to the params exposed by the operation we also need to offer global params because you can apply them whenever.

If we're already in the global scope we can skip this injection step because current scope == global scope.  If we didn't have this logic you'd be offered the same completions twice.

Reply:
That makes sense